src/main/java/core/basesyntax/Car.java

```diff
@@ -1,13 +1,77 @@
 package core.basesyntax;
 
+import java.util.ArrayList;
 import java.util.List;
+import java.util.Objects;
 
-/**
- * Make this class immutable. See requirements in the README file
- */
-public class Car {
-    private int year;
-    private String color;
-    private List<Wheel> wheels;
-    private Engine engine;
+public final class Car {
+    private final int year;
+    private final String color;
+    private final List<Wheel> wheels;
+    private final Engine engine;
+
+    public Car(int year, String color, List<Wheel> wheels, Engine engine) {
+        this.year = year;
+        this.color = color;
+        this.wheels = copyOfWheels(wheels);
+        this.engine = engine == null ? null : engine.clone();
+    }
+
+    public int getYear() {
+        return year;
+    }
+
+    public String getColor() {
+        return color;
+    }
+
+    public List<Wheel> getWheels() {
+        return copyOfWheels(wheels);
+    }
+
+    public Engine getEngine() {
+        return engine == null ? null : engine.clone();
+    }
+
+    public Car changeEngine(Engine engine) {
+        return new Car(year, color, wheels, engine);
+    }
+
+    public final Car changeColor(String newColor) {
+        return new Car(year, newColor, wheels, engine);
+    }
+
+    public final Car addWheel(Wheel newWheel) {
+        Car car = new Car(year, color, wheels, engine);
+        car.wheels.add(newWheel.clone());
+        return car;
+    }
+
+    private List<Wheel> copyOfWheels(List<Wheel> wheelList) {
+        List<Wheel> clone = new ArrayList<>();
```

Comment:
```suggestion
        List<Wheel> clonedWheels = new ArrayList<>(wheelList.size());
```
if capacity will be more than 10, it is good to set it from the beginning. performance wise